packages/flutter/lib/src/material/filled_button.dart

```diff
@@ -0,0 +1,726 @@
+    super.onLongPress,
+    super.onHover,
+    super.onFocusChange,
+    super.style,
+    super.focusNode,
+    bool? autofocus,
+    Clip? clipBehavior,
+    required Widget icon,
+    required Widget label,
+  }) : assert(icon != null),
+       assert(label != null),
+       super(
+         autofocus: autofocus ?? false,
+         clipBehavior: clipBehavior ?? Clip.none,
+         child: _FilledButtonWithIconChild(icon: icon, label: label)
+      );
+
+  _FilledButtonWithIcon.tonal({
+    super.key,
+    required super.onPressed,
+    super.onLongPress,
+    super.onHover,
+    super.onFocusChange,
+    super.style,
+    super.focusNode,
+    bool? autofocus,
+    Clip? clipBehavior,
+    required Widget icon,
+    required Widget label,
+  }) : assert(icon != null),
+       assert(label != null),
+       super.tonal(
+         autofocus: autofocus ?? false,
+         clipBehavior: clipBehavior ?? Clip.none,
+         child: _FilledButtonWithIconChild(icon: icon, label: label)
+       );
+
+  @override
+  ButtonStyle defaultStyleOf(BuildContext context) {
+    final EdgeInsetsGeometry scaledPadding = ButtonStyleButton.scaledPadding(
+      const EdgeInsetsDirectional.fromSTEB(12, 0, 16, 0),
+      const EdgeInsets.symmetric(horizontal: 8),
+      const EdgeInsetsDirectional.fromSTEB(8, 0, 4, 0),
+      MediaQuery.maybeOf(context)?.textScaleFactor ?? 1,
+    );
+    return super.defaultStyleOf(context).copyWith(
+      padding: MaterialStatePropertyAll<EdgeInsetsGeometry>(scaledPadding),
+    );
+  }
+}
+
+class _FilledButtonWithIconChild extends StatelessWidget {
+  const _FilledButtonWithIconChild({ required this.label, required this.icon });
+
+  final Widget label;
+  final Widget icon;
+
+  @override
+  Widget build(BuildContext context) {
+    final double scale = MediaQuery.maybeOf(context)?.textScaleFactor ?? 1;
+    final double gap = scale <= 1 ? 8 : lerpDouble(8, 4, math.min(scale - 1, 1))!;
+    return Row(
+      mainAxisSize: MainAxisSize.min,
+      children: <Widget>[icon, SizedBox(width: gap), Flexible(child: label)],
+    );
+  }
+}
+
+// BEGIN GENERATED TOKEN PROPERTIES - FilledButton
+
+// Do not edit by hand. The code between the "BEGIN GENERATED" and
+// "END GENERATED" comments are generated from data in the Material
+// Design token database by the script:
+//   dev/tools/gen_defaults/bin/gen_defaults.dart.
+
+// Token database version: v0_101
+
+class _FilledButtonDefaultsM3 extends ButtonStyle {
+  _FilledButtonDefaultsM3(this.context)
+   : super(
+       animationDuration: kThemeChangeDuration,
+       enableFeedback: true,
+       alignment: Alignment.center,
+     );
+
+  final BuildContext context;
+  late final ColorScheme _colors = Theme.of(context).colorScheme;
+
+  @override
+  MaterialStateProperty<TextStyle?> get textStyle =>
+    MaterialStatePropertyAll<TextStyle?>(Theme.of(context).textTheme.labelLarge);
+
+  @override
+  MaterialStateProperty<Color?>? get backgroundColor =>
```

Comment:
In methods like this you could avoid creating closures:
```dart
Color? getBackgroundColor(Set<MaterialState> states) {
  if (states.contains(MaterialState.disabled)) {
    return _colors.onSurface.withOpacity(0.12);
  }
  return _colors.primary;
}
MaterialStateProperty<Color?>? get backgroundColor => MaterialStateProperty.resolveWith(getBackgroundColor);
```

Also: shouldn't overrides like this return `MaterialStateProperty<Color?>`, not a nullable MaterialStateProperty?

Reply:
Is the concern here that we are generating short lived objects with the closure (and thus a potential performance issue), or is this just a style preference? I'll change it, but I am just curious.

And yes the override should be `MaterialStateProperty<Color?>?` to be consistent. I think this wasn't flagged as overrides are allowed to constrain the return value like this?

Reply:
Actually, this is turning into a large change, so I will do this closure removal for all the templates in another PR.

Reply:
That sounds good. I suggested it because there should be a small performance benefit.